### PR TITLE
Send more data to Zynq 

### PR DESCRIPTION
**Pull request by pkotamnives**
This PR will resolve a few ZynqMonPath tags in issues. More test and debugging are needed, but the infrastructure is in place. 

The following data streams have been added only for REV2:
- clk monitoring 
- if a 6bit-mask if any 12-ch FFs are 25Gbs ones 
- if a 20bit-mask if any FFs are present 
- optical receiver's power for all receivers/transceivers with 25Gbs 

We also need to sync the data structure in this code with the one in [zynqmon_2](https://github.com/apollo-lhc/sm_cm_config/blob/main/data/zynqmon_2.yml#L2-#L5)

### CODE REVIEW

**Comment by pwittich** [comment]
To shut up codacy maybe add a code check like

```
If ( i > SOME_MAX_VAL) return SOME_ERROR_CONDITION;
```
Where you have to figure out a sensible SOME_MAX_VAL and SOME_ERROR_CONDITION. In particular not obvious to me what the error condition should be.
You have the configASSERT but that is not a standard C thing so I don't think codacy recognizes it.

File: sm_cm_config/src/xml_generate.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+            return True
+        if (self.start <= other.end and self.end >= other.end):
+            return True
+        if (self.start >= other.start and self.end <= other.end):
+            return True
+        return False
+    def overloads(self, other):
+        if (self.start >= 237):
+            return True
+        return False
```
Comment (pwittich) [inline sm_cm_config/src/xml_generate.py:228]: "others" isn't used and so I am not sure that the test on line 254 does what you mean it to do?

Comment (pkotamnives) [reply sm_cm_config/src/xml_generate.py:228]: it was working when I tested it but you're right that this syntax doesn't look good. I'll fix it to :
```
def overloads(self):
    if (self.start >= 237):
        return True
    return False
```

and then just call with `overloads(entries[len(entries)-1])`

Comment (pkotamnives) [reply sm_cm_config/src/xml_generate.py:228]: it was a mistake from copying and pasting...

Comment (pwittich) [reply sm_cm_config/src/xml_generate.py]: ok, I still don't think we should have that number 237 in there, shouldn't it be something like the start + size of the entry compared to the max?

File: projects/cm_mcu/LocalTasks.c [hunk 2]
```diff
@@ -243,11 +249,15 @@ struct MonitorI2CTaskArgs_t ffl12_f1_args = {
...
     .n_values = NSUPPLIES_FFL12_F1 * NPAGES_FFL12_F1 * NCOMMANDS_FFL12_F1,
     .n_pages = NPAGES_FFL12_F1,
+    .selpage_reg = FF_SELPAGE_REG,
     .sm_values = ffl12_f1_values,
     .smbus = &g_sMaster4,
     .smbus_status = &eStatus4,
     .xSem = NULL,
     .ffpart_bit_mask = 0U,
+    .n_rxchs = FF_12_NRXCH,
+    .opt_pow_values = ffl12_f1_optpow_values,
```
Comment (pwittich) [inline projects/cm_mcu/LocalTasks.c:259]: why are these here and not just part of the regular data stored in .sm_values?

File: projects/cm_mcu/LocalTasks.c [hunk 3]
```diff
@@ -281,9 +291,10 @@ struct sm_command_t sm_command_ffldaq_f2[] = {
     {1, 0x00, 0x16, 2, "FF_TEMPERATURE", 0xff, "C", PM_STATUS},
     {1, 0x00, 0x03, 1, "FF_LOS_ALARM", 0xff, "", PM_STATUS},
     {1, 0x00, 0x05, 1, "FF_CDR_LOL_ALARM", 0xff, "", PM_STATUS},
-
+    {2, 0x00, 0x22, 1, "FF_OPT_POW", 0xff, "mw", PM_STATUS}, // read 4 Rx-ch registers with increasing addresses
```
Comment (pwittich) [inline projects/cm_mcu/LocalTasks.c:294]: I was expecting to see an entry for every one of the four channels here, where are they?

File: projects/cm_mcu/LocalTasks.c [hunk 4]
```diff
@@ -319,7 +334,7 @@ struct sm_command_t sm_command_fflot_f2[] = {
     {1, 0x00, 0x16, 2, "FF_TEMPERATURE", 0xff, "C", PM_STATUS},
     {2, 0x00, 0x07, 1, "FF_LOS_ALARM", 0xffff, "", PM_STATUS},
     {2, 0x00, 0x14, 1, "FF_CDR_LOL_ALARM", 0xffff, "", PM_STATUS},
-
+    {2, 0x01, 0xe4, 1, "FF_OPT_POW", 0xff, "mw", PM_STATUS}, // read 12 Rx-ch registers  with decreasing addresses
```
Comment (pwittich) [inline projects/cm_mcu/LocalTasks.c:337]: same question

File: projects/cm_mcu/LocalTasks.c (hunk 2 above)
Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c:259]: From my recollection, I thought that we would send 4 values per for each 4-ch XCVR FFs and 12 values for each 12-ch Rx FFs so I separate this out from .sm_values that has a one-to-one device to register output. At this point, we store an average value from 4-ch/12-ch FFs. If you suggest removing this .opt_pow_value and add its values to .sm_values, I can do that too. 

However, for future development (in case the cm addresses in Zynq are increased), we can have this separate attribute  come into handy?

File: projects/cm_mcu/LocalTasks.c (hunk 3 above)
Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c:294]: As I have mentioned, we store an average optical power per FF with 25Gbs. The magic of reading each channel is [here](https://github.com/apollo-lhc/cm_mcu/blob/b6b3539f78c9617dd99e43e54c2f3551ee39504e/projects/cm_mcu/MonitorI2CTask.c#L228-L232).

File: sm_cm_config/src/xml_generate.py (hunk 1 above)
Comment (pkotamnives) [reply sm_cm_config/src/xml_generate.py]: This 237 was a pure experimental value. I tried a few different starting addresses (even though start + size < 256) of the last node but ran into a wrap-around issue until I found this value. So, we can think of this number as a safe maximum starting value for now?

File: projects/cm_mcu/LocalTasks.c (hunk 3 above)
Comment (pwittich) [reply projects/cm_mcu/LocalTasks.c:294]: do we have to use magic? I thought we would just enumerate the registers we want to read, no magic required.

File: projects/cm_mcu/LocalTasks.c (hunk 2 above)
Comment (pwittich) [reply projects/cm_mcu/LocalTasks.c:259]: we are starting to treat this struct more and more like a grab-bag with slots of entries that only count for one device or another, and the code is full of exceptions which will make it hard to debug and extend.

File: projects/cm_mcu/LocalTasks.c (hunk 3 above)
Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c:294]: I understand your point that the registers can be explicitly added here. Perhaps you are suggesting that we should change the CLI command for reading optical power to shows this value in all channels? but for now we can only send an average value to zynq, right?

File: projects/cm_mcu/LocalTasks.c (hunk 2 above)
Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c:259]: Okay, I get your point. I will add optical power values back to .sm_values and remove this attribute.

File: projects/cm_mcu/LocalTasks.c (hunk 4 above)
Comment (pwittich) [reply projects/cm_mcu/LocalTasks.c:337]: another way to solve this problem is to use the 4th field which I think is the size of the data and make it something like 4 for the 4 channel devices and 12 for the 12 channel Rx device... (double check if that is the meaning of the 4th field ;-) )

File: sm_cm_config/src/xml_generate.py (hunk 1 above)
Comment (pwittich) [reply sm_cm_config/src/xml_generate.py]: I'm not sure I understand what this is supposed to do and why if start + size < 256 it would create a problem? I'm a bit uncomfortable with this approach because it sounds like we don't actually understand what is going on.

File: projects/cm_mcu/LocalTasks.c (hunk 3 above)
Comment (pwittich) [reply projects/cm_mcu/LocalTasks.c:294]: Dan is going to change the data size so we can send the full information soon ... so we should be ready for that. I guess we can merge this PR with just the average value for sending.

File: projects/cm_mcu/LocalTasks.c (hunk 4 above)
Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c]: I was not sure what you meant by "the 4th field." Perhaps you meant sending output from every fourth channel? So, we would send 8 outputs for XCVR devices and 12 outputs for Rx devices.

File: sm_cm_config/src/xml_generate.py (hunk 1 above)
Comment (pkotamnives) [reply sm_cm_config/src/xml_generate.py]: I honestly did not understand why the wrap-around could have caused when the last address is still within 256. I can demonstrate the issue again. However, since we will have more address space, I believe that we will come back to this script for future development.

Comment (pkotamnives) [reply sm_cm_config/src/xml_generate.py]: when xml file has a start node < 237 (shows an expected table)
```
CM_FFARGV_MON| IS_FF25GBS| IS_PRESENT|
               --|-----------|-----------|
       F1_FFL12CH|          0|         63|
        F1_FFLDAQ|         15|         15|
       F2_FFL12CH|          0|         63|
        F2_FFLDAQ|         15|         15|
```
when xml file has start node > 237 (shows unexpected table)
```
CM_FFARGV_MON| IS_FF25GBS| IS_PRESENT|
               --|-----------|-----------|
       F1_FFL12CH|         15|         15|
        F1_FFLDAQ|          0|         63|
       F2_FFL12CH|         15|         15|
        F2_FFLDAQ|         63|         15|
```
For the starting node, I meant this [line](https://github.com/apollo-lhc/sm_cm_config/blob/main/data/zynqmon_2.yml#L217)

File: projects/cm_mcu/InitTask.c [hunk 5]
```diff
@@ -55,8 +56,25 @@ void InitTask(void *parameters)
...
+  char progeeprom_temp[5][CLOCK_EEPROM_PROGNAME_REG_NAME];
   for (int i = 0; i < 5; ++i) {
     init_load_clk(i); // load each clock config from EEPROM
+    // get and print out the file name
+    vTaskDelay(pdMS_TO_TICKS(500));
+    char progname_clkdesgid[CLOCK_PROGNAME_REG_NAME];     // program name from DESIGN_ID register of clock chip
+    char progname_eeprom[CLOCK_EEPROM_PROGNAME_REG_NAME]; // program name from eeprom
+    getClockProgram(i, progname_clkdesgid, progname_eeprom);
+    strcpy(progchip_temp[i], progname_clkdesgid);
+    strcpy(progeeprom_temp[i], progname_eeprom);
```
Comment (pwittich) [inline projects/cm_mcu/InitTask.c:69]: use `strncpy` instead to avoid buffer overruns

Comment (pkotamnives) [reply projects/cm_mcu/InitTask.c]: thank you. Done.

File: projects/cm_mcu/clocksynth.c [hunk 6]
```diff
@@ -74,10 +74,10 @@ void getClockProgram(int device, char progname_clkdesgid[CLOCK_PROGNAME_REG_NAME
     return;
 
   // grab the semaphore to ensure unique access to I2C controller
```
Comment (pwittich) [inline projects/cm_mcu/clocksynth.c:76]: why is this commented out?

File: projects/cm_mcu/Tasks.h [hunk 7]
```diff
@@ -170,6 +170,8 @@ struct dev_moni2c_addr_t {
   uint8_t mux_bit;              // port of the mux; write value 0x1U<<mux_bit to the mux register
   uint8_t dev_addr;             // I2C address of device.
   uint16_t eeprom_progname_reg; // register on eeprom for reading program version
+  const char *configname_chip;
+  const char *configname_eeprom;
```
Comment (pwittich) [inline projects/cm_mcu/Tasks.h:174]: these three variables are not generic for all devices, are they? they are just for clock devices. Should they be stored elsewhere?

File: projects/cm_mcu/LocalTasks.c [hunk 8]
```diff
@@ -195,6 +200,21 @@ struct sm_command_t sm_command_fflit_f1[] = {
...
+    {1, 0x00, 0x00, 1, "FF_CH03_OPT_POW", 0xff, "mw", PM_STATUS},
+    {1, 0x00, 0x00, 1, "FF_CH04_OPT_POW", 0xff, "mw", PM_STATUS},
+    {1, 0x00, 0x00, 1, "FF_CH05_OPT_POW", 0xff, "mw", PM_STATUS},
+    {1, 0x00, 0x00, 1, "FF_CH06_OPT_POW", 0xff, "mw", PM_STATUS},
+    {1, 0x00, 0x00, 1, "FF_CH07_OPT_POW", 0xff, "mw", PM_STATUS},
+    {1, 0x00, 0x00, 1, "FF_CH08_OPT_POW", 0xff, "mw", PM_STATUS},
+    {1, 0x00, 0x00, 1, "FF_CH09_OPT_POW", 0xff, "mw", PM_STATUS},
+    {1, 0x00, 0x00, 1, "FF_CH10_OPT_POW", 0xff, "mw", PM_STATUS},
+    {1, 0x00, 0x00, 1, "FF_CH11_OPT_POW", 0xff, "mw", PM_STATUS},
+    {1, 0x00, 0x00, 1, "FF_CH12_OPT_POW", 0xff, "mw", PM_STATUS},
```
Comment (pwittich) [inline projects/cm_mcu/LocalTasks.c:217]: i understand why you did it this way, but we should think of a better solution than this to solve this problem

File: projects/cm_mcu/LocalTasks.c [hunk 9]
```diff
@@ -612,28 +687,96 @@ TickType_t getFFupdateTick(int mask)
...
+    }
+    avg_val = sum_val / (ffl12_f2_args.n_commands - 4);
+  }
+  else {
+    for (int i1 = 4; i1 < ffldaq_f2_args.n_commands; ++i1) {
+      int dev = i - ff_moni2c_arg[3].int_idx + ff_moni2c_arg[3].dev_int_idx;
+      int index = (dev) * (ffldaq_f2_args.n_commands * ffldaq_f2_args.n_pages) + i1;
+      sum_val += ffldaq_f2_args.sm_values[index];
+    }
+    avg_val = sum_val / (ffldaq_f2_args.n_commands - 4);
```
Comment (pwittich) [inline projects/cm_mcu/LocalTasks.c:755]: a lot of copypasta here. can you clean this up by maybe usinga  pointer for the ff*args in the repeated code, and just setting the appropriate pointer 1x?

File: projects/cm_mcu/LocalTasks.c [hunk 10]
```diff
@@ -612,28 +687,96 @@ TickType_t getFFupdateTick(int mask)
...
+
+  else if (i == 1) {
+    val = ffldaq_f1_args.present_bit_mask;
+  }
+
+  else if (i == 2) {
+    val = ffl12_f2_args.present_bit_mask;
+  }
+  else {
+    val = ffldaq_f2_args.present_bit_mask;
```
Comment (pwittich) [inline projects/cm_mcu/LocalTasks.c:779]: ditto, cleaner to select the appropriate ff*args and the have the assignment 1x.

File: projects/cm_mcu/MonitorI2CTask.h [hunk 11]
```diff
@@ -35,46 +35,51 @@ struct MonitorI2CTaskArgs_t {
...
   const int n_values;                // number of results
   const uint8_t n_pages;             // number of pages to loop over
+  const uint16_t selpage_reg;        // register for selecting page
   uint16_t *sm_values;
   tSMBus *smbus;                       // pointer to I2C controller
   volatile tSMBusStatus *smbus_status; // pointer to I2C status
   TickType_t updateTick;               // last update time, in ticks
   SemaphoreHandle_t xSem;              // semaphore for controlling access to device
   uint8_t ffpart_bit_mask;             // this mask is only used for detecting 12-ch 25Gbps on the REV2 board
+  uint32_t present_bit_mask;           // this mask is used for all ffs to detect if it is mounted or not
```
Comment (pwittich) [inline projects/cm_mcu/MonitorI2CTask.h:45]: not clear if these belong here, in the argument to the monitor task.

File: projects/cm_mcu/ZynqMonTask.c [hunk 12]
```diff
@@ -50,16 +52,16 @@
...
+  message[1] |= ((sensor & 0x3) << 4) | ((data >> 12) & 0xF); // 0b00 dddddd
   // data frame 2, data[11:6] (6 bits)
   message[2] = SENSOR_MESSAGE_DATA_FRAME;
-  message[2] |= (data >> 6) & 0x3F;
+  message[2] |= (data >> 6) & 0x3F; // 0b00 dddddd
   // data frame 3, data[5:0] ( 6 bits )
   message[3] = SENSOR_MESSAGE_DATA_FRAME;
-  message[3] |= data & 0x3F;
+  message[3] |= data & 0x3F; // 0b00 dddddd
 }
```
Comment (pwittich) [inline projects/cm_mcu/ZynqMonTask.c:65]: what do these comments mean?

File: projects/cm_mcu/ZynqMonTask.c [hunk 13]
```diff
@@ -268,17 +270,123 @@ void zm_set_firefly_temps(struct zynqmon_data_t data[], int start)
...
+    uint16_t val;
+    if (isFFStale()) {
+      data[ll].data.us = 0xFF; // special stale value
+    }
+    else {
+      if (j == 0)
+        val = ((ffl12_f1_args.ffpart_bit_mask & 0x1) << 1) | (ffl12_f1_args.ffpart_bit_mask & 0x1) | ((ffl12_f1_args.ffpart_bit_mask & 0x2) << 2) | ((ffl12_f1_args.ffpart_bit_mask & 0x2) << 1) | ((ffl12_f1_args.ffpart_bit_mask & 0x4) << 3) | ((ffl12_f1_args.ffpart_bit_mask & 0x4) << 2);
+      else if (j == 2)
+        val = ((ffl12_f2_args.ffpart_bit_mask & 0x1) << 1) | (ffl12_f2_args.ffpart_bit_mask & 0x1) | ((ffl12_f2_args.ffpart_bit_mask & 0x2) << 2) | ((ffl12_f2_args.ffpart_bit_mask & 0x2) << 1) | ((ffl12_f2_args.ffpart_bit_mask & 0x4) << 3) | ((ffl12_f2_args.ffpart_bit_mask & 0x4) << 2);
+      else
```
Comment (pwittich) [inline projects/cm_mcu/ZynqMonTask.c:302]: this code might need a comment or two to explain what it is doing. also a lot of magic numbers?

File: projects/cm_mcu/ZynqMonTask.c [hunk 14]
```diff
@@ -380,41 +520,9 @@ void zm_set_fpga(struct zynqmon_data_t data[], int start)
...
-  zm_set_adcmon(&zynqmon_data[104], 128);
-  // uptime, size 2
-  zm_set_uptime(&zynqmon_data[125], 192);
-  // gitversion, size 20
-  zm_set_gitversion(&zynqmon_data[127], 118);
-  // fpga, size 8
-  zm_set_fpga(&zynqmon_data[137], 150);
-}
-#define ZMON_VALID_ENTRIES 145
+#define ZMON_VALID_ENTRIES 233
```
Comment (pwittich) [inline projects/cm_mcu/ZynqMonTask.c:525]: aren't these now set in `Zynqmon_Addresses.c`?

File: projects/cm_mcu/LocalTasks.c (hunk 4 above)
Comment (pwittich) [reply projects/cm_mcu/LocalTasks.c]: I mean the fourth field in each row, i.e., the `size` entry. see here: https://github.com/apollo-lhc/cm_mcu/blob/fd4fb8085d121b84480b9cb9605624a8439b3e14/projects/cm_mcu/MonitorI2CTask.h#L16-L20

Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c]: the issue of reading 4 or 12 bytes is that the register addresses for the 12-ch one are decreasing but the default i2c operation would read increasing register addresses. Though we can reorder outputs for 12-ch devices but this will add on to another exception.

File: projects/cm_mcu/clocksynth.c (hunk 6 above)
Comment (pkotamnives) [reply projects/cm_mcu/clocksynth.c:76]: there is already semaphore take in a larger scope where `getClockProgram` is called. See https://github.com/apollo-lhc/cm_mcu/blob/78ae126eeb029e2c4a4f7a6b1d28e2ab8be7bde0/projects/cm_mcu/InitTask.c#L57

File: projects/cm_mcu/Tasks.h (hunk 7 above)
Comment (pkotamnives) [reply projects/cm_mcu/Tasks.h:174]: You're right. I think having this information associated with a device struct is quite handy in a few places `projects/cm_mcu/commands/SensorControl.c`, `projects/cm_mcu/InitTask.c`, and `projects/cm_mcu/clocksynth.c` which need to access it and update/read. Unless you have other suggestions?

File: projects/cm_mcu/LocalTasks.c (hunk 8 above)
Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c:217]: the issue has something to do with the size of commands and outputs that must be declared as in [here](https://github.com/apollo-lhc/cm_mcu/blob/78ae126eeb029e2c4a4f7a6b1d28e2ab8be7bde0/projects/cm_mcu/LocalTasks.c#L275-L276). So, in order to keep the same size, I needed to create a set of fake addresses.

File: projects/cm_mcu/ZynqMonTask.c (hunk 12 above)
Comment (pkotamnives) [reply projects/cm_mcu/ZynqMonTask.c:65]: it was a note to myself what the encoder does at the beginning of this work. I'll remove it

File: projects/cm_mcu/ZynqMonTask.c (hunk 14 above)
Comment (pkotamnives) [reply projects/cm_mcu/ZynqMonTask.c:525]: you're right. However, I got a compile error for not declaring them in this file. I think some lines in the same file such as [this](https://github.com/apollo-lhc/cm_mcu/blob/78ae126eeb029e2c4a4f7a6b1d28e2ab8be7bde0/projects/cm_mcu/ZynqMonTask.c#L531) rely on them. 

Should we declare these two variables in Task.h? instead of both ZynMon*.cc

File: projects/cm_mcu/LocalTasks.c (hunk 8 above)
Comment (pwittich) [reply projects/cm_mcu/LocalTasks.c:217]: Right, I get it, but I think we should figure out something better, if not in this PR, then later.

File: projects/cm_mcu/ZynqMonTask.c (hunk 14 above)
Comment (pwittich) [reply projects/cm_mcu/ZynqMonTask.c:525]: Maybe we should instead pass it in as an argument to the `zm_send_data`? That is probably cleaner anyhow.

File: projects/cm_mcu/Tasks.h (hunk 7 above)
Comment (pwittich) [reply projects/cm_mcu/Tasks.h:174]: Maybe we should create a way to store them

File: projects/cm_mcu/clocksynth.c (hunk 6 above)
Comment (pwittich) [reply projects/cm_mcu/clocksynth.c:76]: Ok, then please remove it and add a comment at the top of the function that you _must_ grab the semaphore before calling this function

File: projects/cm_mcu/ZynqMonTask.c (hunk 14 above)
Comment (pwittich) [reply projects/cm_mcu/ZynqMonTask.c:525]: i.e., modify the `mcu_generate.py` to make the function `void zm_fill_structs(int valid_entries)`

Comment (pwittich) [reply projects/cm_mcu/ZynqMonTask.c:525]: actually I guess that would not solve the problem.

File: projects/cm_mcu/clocksynth.c (hunk 6 above)
Comment (pkotamnives) [reply projects/cm_mcu/clocksynth.c:76]: Done.

File: projects/cm_mcu/LocalTasks.c (hunk 10 above)
Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c]: Done.

File: projects/cm_mcu/LocalTasks.c (hunk 9 above)
Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c]: Done

File: projects/cm_mcu/MonitorI2CTask.h (hunk 11 above)
Comment (pkotamnives) [reply projects/cm_mcu/MonitorI2CTask.h:45]: I think you're right. It doesn't seem to be necessary to have this information in a moni2c struct. I will remove it.

Comment (pkotamnives) [reply projects/cm_mcu/MonitorI2CTask.h:45]: Yes, it's part of this struct. It's necessary for [this function](https://github.com/apollo-lhc/cm_mcu/blob/77ff48fb0ce36ae73250b3790da4cc1571f4b68a/projects/cm_mcu/LocalTasks.c#L738-L747).

File: projects/cm_mcu/ZynqMonTask.c (hunk 13 above)
Comment (pkotamnives) [reply projects/cm_mcu/ZynqMonTask.c]: Done.

File: projects/cm_mcu/Tasks.h (hunk 7 above)
Comment (pkotamnives) [reply projects/cm_mcu/Tasks.h:174]: They are varying by clock chips so I thought it made sense to have them be an attribute in a struct. Though, it's not clean for other non-clocks. Do you have an alternative way to store them?

Comment (pwittich) [reply projects/cm_mcu/Tasks.h:174]: also there is no storage associated with this and no variables in global scope that this points to. Instead we point to some variables on the heap in `InitTask`. Maybe it makes sense to instead create a separate struct somewhere, with storage associated to it, and then put them in there.
If `InitTask` ever goes out of scope the pointers you set there will point to random addresses and contain possibly garbage.

File: projects/cm_mcu/InitTask.c [hunk 15]
```diff
@@ -55,8 +56,25 @@ void InitTask(void *parameters)
...
+    strncpy(progeeprom_temp[i], progname_eeprom, 20);
+    if (i == 0) {
+      clkr0a_moni2c_addrs[0].configname_chip = progchip_temp[i];
+      clkr0a_moni2c_addrs[0].configname_eeprom = progeeprom_temp[i];
+    }
+    else {
+      clk_moni2c_addrs[i - 1].configname_chip = progchip_temp[i];
+      clk_moni2c_addrs[i - 1].configname_eeprom = progeeprom_temp[i];
+    }
   }
```
Comment (pwittich) [inline projects/cm_mcu/InitTask.c:78]: This is what I was referring to in the other comment. the `_temp` variables are in local scope in the `InitTask`. If or when the `vTaskDelete` at the end of the inittask gets called, the `_temp` variables are no longer around.

File: projects/cm_mcu/MonitorI2CTask.h (hunk 11 above)
Comment (pwittich) [reply projects/cm_mcu/MonitorI2CTask.h:45]: i don't doubt we have a reason to store it. I just don't think it belongs in this struct. Same comment as in other places.

Comment (pkotamnives) [reply projects/cm_mcu/MonitorI2CTask.h]: Done.

File: projects/cm_mcu/InitTask.c (hunk 15 above)
Comment (pkotamnives) [reply projects/cm_mcu/InitTask.c]: Done.

File: projects/cm_mcu/Tasks.h (hunk 7 above)
Comment (pkotamnives) [reply projects/cm_mcu/Tasks.h]: Done.

File: projects/cm_mcu/LocalTasks.c (hunk 8 above)
Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c:223]: Noted, this will be in a new PR.
